stan/math/opencl/kernel_generator/operation.hpp

```diff
@@ -0,0 +1,226 @@
+   * @tparam T_lhs type of the left-hand-side expression
+   * @param lhs Left-hand-side expression
+   * @return kernel source
+   */
+  template <typename T_lhs>
+  inline std::string get_kernel_source_for_evaluating_into(
+      const T_lhs& lhs) const;
+
+  template <typename T_lhs>
+  struct cache {
+    static cl::Kernel kernel;  // cached kernel - different for every
+                               // combination of template instantination of \c
+                               // operation and every \c T_lhs
+  };
+
+  /**
+   * generates kernel code for this and nested expressions.
+   * @param[in,out] generated set of (pointer to) already generated operations
+   * @param name_gen name generator for this kernel
+   * @param i row index variable name
+   * @param j column index variable name
+   * @return part of kernel with code for this and nested expressions
+   */
+  inline kernel_parts get_kernel_parts(std::set<const void*>& generated,
+                                       name_generator& name_gen,
+                                       const std::string& i,
+                                       const std::string& j) const {
+    kernel_parts res{};
+    if (generated.count(this) == 0) {
+      generated.insert(this);
+      std::array<kernel_parts, N> args_parts = index_apply<N>([&](auto... Is) {
+        return std::array<kernel_parts, N>{
+            std::get<Is>(arguments_)
+                .get_kernel_parts(generated, name_gen, i, j)...};
+      });
+      res.body
+          = std::accumulate(args_parts.begin(), args_parts.end(), std::string(),
+                            [](const std::string& a, const kernel_parts& b) {
+                              return a + b.body;
+                            });
+      res.args
+          = std::accumulate(args_parts.begin(), args_parts.end(), std::string(),
+                            [](const std::string& a, const kernel_parts& b) {
+                              return a + b.args;
+                            });
+      this->var_name = name_gen.generate();
+      kernel_parts my_part = index_apply<N>([&](auto... Is) {
+        return this->derived().generate(i, j,
+                                        std::get<Is>(arguments_).var_name...);
+      });
+      res.body += my_part.body;
+      res.args += my_part.args;
+    }
+    return res;
+  }
+
+  /**
+   * Sets kernel arguments for nested expressions.
+   * @param[in,out] generated set of expressions that already set their kernel
+   * arguments
+   * @param kernel kernel to set arguments on
+   * @param[in,out] arg_num consecutive number of the first argument to set.
+   * This is incremented for each argument set by this function.
+   */
+  inline void set_args(std::set<const void*>& generated, cl::Kernel& kernel,
+                       int& arg_num) const {
+    if (generated.count(this) == 0) {
+      generated.insert(this);
+      index_apply<N>([&](auto... Is) {
+        (void)std::initializer_list<int>{
```

Comment:
It is strictly speeking not needed. But it prevents compiler throwing warnings about unused expressions. We are only constructing this initializer list so we can do the parameter pack expansion. Expansion returns a comma-separated list of values, which can not be used as a standalone expression. Since `set_args()` returns `void` we do a trick with a comma operator to have get int values and construct an initializer list from them. I guess I should add a comment in code explaining this.